### ISSUE TITLE
Warning: a promise was created in a  handler but was not returned from it

### ISSUE THREAD

**Issue by luknei** [petkaantonov/bluebird#955]
I know similar issues have been created in the past. But after reading them I still cannot figure out a right solution how to go around this warning in my situation. 

So the situation is that I have a HttpRequest class that has constructor. At the time of instantiation there is no response yet, but I want it to be there as a promise of the future value. The response promise is being created after the method send is called. However, there are places where I want to subscribe to HttpRequest response property when send method is still not called. As far as I understand promises it seems logical (please correct me if I'm wrong). However I still get that annoying warning that "a promise was created in a  handler but was not returned from".

Thanks,
Lukas

```javascript
export default class HttpRequest {
    constructor(router) {
        this._router = router;
        this._fulfull = null;
        this._reject = null;

        this.response = new Promise((fulfill, reject) => {
            this._fulfill = fulfill;
            this._reject = reject;
        });
    }

    send() {
        let response = this._router.send(this);

        response.then(this._fulfull, this._reject);

        return this.response;
    }
}
```

**Comment by nathanmarks** [comment]
```js
  send() {
    return this._router.send(this).then(
      this._fulfull,
      this._reject
    ).then(() => this.response);
  }
```

**Comment by luknei** [comment]
@nathanmarks Thanks for your reply! That's really good point, I'd say  - promises 101 :) However that is not what triggers that error :/

**Comment by nathanmarks** [comment]
@luknei Also, I'm pretty sure that the way you are using this falls under the deferred anti-pattern. It seems like a poor use of promises IMO and is hard to follow.

Why don't you use something like `EventEmitter` ([docs](https://nodejs.org/api/events.html#events_class_eventemitter)) to notify the other places that something has happened and/or is ready.

**Comment by luknei** [comment]
@nathanmarks Well I think promise in this case suits perfectly. From the logical point of view the response is a placeholder for a future value that may or may not be resolved some time in the future at most - once. Moreover, request cannot have a multiple response or stream of responses. So as far as I know it's perfectly fine to use promises for such reasons as completion event.

And according to MDN:
> The Promise object is used for deferred and asynchronous computations. A Promise represents an operation that hasn't completed yet, but is expected in the future.

So in this case whenever HttpRequest is created is logical to expect that request eventually will have a response.
For example - the promise could be rejected in case of error response, or request taking too long or not sending request in the first place and thus making it more easy to track down such situations in the code.

**Comment by nathanmarks** [comment]
@luknei Others probably can weigh on this better than myself -- but notice this line:

>  A Promise represents an operation that hasn't completed yet

In your code there is no operation, it's a promise with nothing. If you did a static analysis on the code there would be no resolution and no operation.

Check out here too: https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern

> Almost a sure sign of using promises as glorified callbacks. Instead of doThat(function(err, success)) you do doThat().then(success, err) and rationalize to yourself that at least the code is "less coupled" or something.

**Comment by petkaantonov** [comment]
Please ask support questions on stack overflow. And attach the stack trace there.
